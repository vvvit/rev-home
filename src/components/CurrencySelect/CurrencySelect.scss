@import '../../styles/variables.scss';

.CurrencySelect {
    position: relative;
    font-size: $font-size-xl;

    &-Value {
        cursor: pointer;
        display: flex;
        align-items: center;
        justify-content: space-between;
        width: 120px;
    }

    &-List {
        border: 1px solid $color-gray;
        border-bottom: 0;
        background: $color-white;
        position: absolute;
        z-index: 1;
        left: -5px;
        top: -5px;
    }

    &-ListItem {
        padding: 4px;
        border-bottom: 1px solid $color-gray;
        cursor: pointer;

        &_selected, &:hover {
            background: $color-light-gray;
        }
    }
}
