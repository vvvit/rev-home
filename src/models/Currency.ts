export const enum CurrencyName {
    USD = 'usd',
    EUR = 'eur',
    GBP = 'gbp',
}

export interface ICurrency {
    name: CurrencyName,
    value: number
}
