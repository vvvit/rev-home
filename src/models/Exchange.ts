export const enum CurrencyExchangeType {
    BUY = 'buy',
    SELL = 'sell'
}
