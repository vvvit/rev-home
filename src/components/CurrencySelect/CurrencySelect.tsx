import * as React from 'react';
import { boundMethod } from 'autobind-decorator';
import { cn } from '@bem-react/classname';

import { CurrencyName } from '../../models/Currency';
import { Icon } from '../Icon/Icon';

import './CurrencySelect.scss';

const cnApp = cn('CurrencySelect');

export interface CurrencySelectProps {
    value: CurrencyName;
    items: CurrencyName[];
    onChangeValue: (value: CurrencyName) => void;
}

interface CurrencySelectState {
    isOpen: boolean;
}

class CurrencySelect extends React.Component<CurrencySelectProps, CurrencySelectState> {
    state: CurrencySelectState = {
        isOpen: false
    };

    render() {
        const { value, items } = this.props;
        const { isOpen } = this.state;

        return (
            <div className={cnApp()}>
                <div onClick={this.toggleSelect} className={cnApp('Value')}>
                    {value}<Icon type={isOpen ? 'arrow-up' : 'arrow-down'} />
                </div>
                {isOpen && (
                    <div className={cnApp('List')}>
                        <div onClick={this.toggleSelect} className={cnApp('ListItem', { selected: true })}>{value}</div>
                        {items
                            .filter(name => name !== value)
                            .map(name => (
                            <div
                                key={name}
                                className={cnApp('ListItem')}
                                onClick={this.createOnSelectItem(name)}
                            >
                                {name}
                            </div>
                        ))}
                    </div>
                )}
            </div>
        );
    }

    @boundMethod
    toggleSelect() {
        this.setState({ isOpen: !this.state.isOpen });
    }

    createOnSelectItem(name: CurrencyName) {
        return () => {
            this.setState({ isOpen: false });
            this.props.onChangeValue(name);
        };
    }
}

export { CurrencySelect };
