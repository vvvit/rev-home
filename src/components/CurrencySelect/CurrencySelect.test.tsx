import * as React from 'react';
import { shallow } from 'enzyme';

import { CurrencySelect } from './CurrencySelect';
import { CurrencyName } from '../../models/Currency';

describe('CurrencySelect', () => {
    let onChange: jest.Mock;

    beforeEach(() => {
        onChange = jest.fn();
    });

    it('renders correctly with EUR value', () => {
        expect(shallow(
            <CurrencySelect
                value={CurrencyName.EUR}
                items={[CurrencyName.EUR, CurrencyName.USD, CurrencyName.RUB]}
                onChangeValue={onChange}
            />
        )).toMatchSnapshot();
    });

    it('correct renders with open state', () => {
        const component = shallow(
            <CurrencySelect
                value={CurrencyName.EUR}
                items={[CurrencyName.EUR, CurrencyName.USD, CurrencyName.RUB]}
                onChangeValue={onChange}
            />
        );

        component.setState({ isOpen: true });

        expect(component).toMatchSnapshot();
    });

    it('correct toggles open state', () => {
        const component = shallow(
            <CurrencySelect
                value={CurrencyName.EUR}
                items={[CurrencyName.EUR, CurrencyName.USD, CurrencyName.RUB]}
                onChangeValue={onChange}
            />
        );

        component.find('.CurrencySelect-Value').simulate('click');

        expect(component.state('isOpen')).toEqual(true);
    });

    it('correct toggles open state', () => {
        const component = shallow(
            <CurrencySelect
                value={CurrencyName.EUR}
                items={[CurrencyName.EUR, CurrencyName.USD, CurrencyName.RUB]}
                onChangeValue={onChange}
            />
        );

        component.find('.CurrencySelect-Value').simulate('click');

        expect(component.state('isOpen')).toEqual(true);
    });

    it('correct toggles open state', () => {
        const component = shallow(
            <CurrencySelect
                value={CurrencyName.EUR}
                items={[CurrencyName.EUR, CurrencyName.USD, CurrencyName.RUB]}
                onChangeValue={onChange}
            />
        );

        component.find('.CurrencySelect-Value').simulate('click');

        expect(component.state('isOpen')).toEqual(true);
    });

    it('correct toggles open state', () => {
        const component = shallow(
            <CurrencySelect
                value={CurrencyName.EUR}
                items={[CurrencyName.EUR, CurrencyName.USD, CurrencyName.RUB]}
                onChangeValue={onChange}
            />
        );

        component.find('.CurrencySelect-Value').simulate('click');

        expect(component.state('isOpen')).toEqual(true);
    });

    it('invokes onchange when item was selected', () => {
        const component = shallow(
            <CurrencySelect
                value={CurrencyName.EUR}
                items={[CurrencyName.EUR, CurrencyName.USD, CurrencyName.RUB]}
                onChangeValue={onChange}
            />
        );
        component.setState({ isOpen: true });
        component.find('.CurrencySelect-ListItem').at(1).simulate('click');

        expect(onChange).toBeCalled();
    });
});
