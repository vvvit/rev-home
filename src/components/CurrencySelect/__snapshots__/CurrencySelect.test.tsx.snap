// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencySelect correct renders with open state 1`] = `
<div
  className="CurrencySelect"
>
  <div
    className="CurrencySelect-Value"
    onClick={[Function]}
  >
    EUR
    <Icon
      size="m"
      type="arrow-up"
    />
  </div>
  <div
    className="CurrencySelect-List"
  >
    <div
      className="CurrencySelect-ListItem CurrencySelect-ListItem_selected"
      onClick={[Function]}
    >
      EUR
    </div>
    <div
      className="CurrencySelect-ListItem"
      key="USD"
      onClick={[Function]}
    >
      USD
    </div>
    <div
      className="CurrencySelect-ListItem"
      key="RUB"
      onClick={[Function]}
    >
      RUB
    </div>
  </div>
</div>
`;

exports[`CurrencySelect renders correctly with EUR value 1`] = `
<div
  className="CurrencySelect"
>
  <div
    className="CurrencySelect-Value"
    onClick={[Function]}
  >
    EUR
    <Icon
      size="m"
      type="arrow-down"
    />
  </div>
</div>
`;
